src/utils.jl

```diff
@@ -458,16 +458,46 @@ function flatten1(ex)
   for x in ex.args
     isexpr(x, :block) ? append!(ex′.args, x.args) : push!(ex′.args, x)
   end
-  # Don't use `unblock` to preserve line nos
+  # Don't use `unblock` to preserve line nos.
   return length(ex′.args) == 1 ? ex′.args[1] : ex′
 end
 
+false_or_bflatten(x) = x==false ? false : bflatten(x)
+
+function bflatten(x)
+    fx = flatten(x)
+    return isexpr(fx, :block) ? fx : Expr(:block,fx)
+end
+
 """
     flatten(ex)
 
 Flatten any redundant blocks into a single block, over the whole expression.
 """
-flatten(ex) = postwalk(flatten1, ex)
+function flatten end
+
+flatten(x) = x
+function flatten(x::Expr)
+  if isexpr(x, :try) # try args can be either false or blocks, but not non-block exprs, so we need to be a bit verbose here
+    isa(x.args[2], Symbol) || x.args[2] == false || error("MacroTools.flatten has encountered a misformed `try` block. Please report this as a bug.")
+    if length(x.args) == 3
+      # try catch end # 3 args, 2nd is a symbol or false, 3rd is block
+      return Expr(x.head, bflatten(x.args[1]), x.args[2], bflatten(x.args[3]))
+    elseif length(x.args) == 4
+      # try finally end # 4 args, 2nd==3rd==false, 4th is block
+      # try catch finally end # 4 args, 2nd is a symbol or false, 3rd and 4th are block
+      return Expr(x.head, bflatten(x.args[1]), x.args[2], false_or_bflatten(x.args[3]), bflatten(x.args[4]))
+    elseif length(x.args) == 5
+      # try catch else end # 5 args, 2nd is a symbol or false, 3rd is block, 4th is false, 5th is block
+      # try catch else finally end # 5 args, 2nd is a symbol or false, the rest are block
+      return Expr(x.head, bflatten(x.args[1]), x.args[2], bflatten(x.args[3]), false_or_bflatten(x.args[4]), bflatten(x.args[5]))
+    else
+      error("MacroTools.flatten has encountered a misformed `try` block. Please report this as a bug.")
+    end
```

Comment:
```suggestion
    else
        return Expr(x.head, map(flatten_args, x.args)...)
    end
```
I think this is very appealing in simplicity --- if it works! It's also easy to explain, `flatten flattens blocks, except those in try clauses, because they have a special meaning`. Yes, technically we can flatten some more in some cases, but I don't think that there's any code out there which relies on this fact, but which isn't also susceptible to the bugs we've been discussing.

Reply:
I think the requirement for try arguments to be either blocks or false (discussed in the comment above) precludes this nicer more elegant treatment.

Reply:
`flatten_args(false)` is `false`, and `flatten_args(some_block)` is a block. Why wouldn't it work?

Reply:
You are right, that would work, but I think this would not be flattened:

```
try
    begin
        f()
        g()
    end
catch
end
```

while at least ResumableFunctions expects this to be flattened.

I implemented what you suggested in the last few minutes and that is how I found the issue.

Reply:
True. Maybe `return Expr(:try, flatten(x.args[1]), map(flatten_args, x.args[2:end])...)` ? But then, that is also problematic in the other arguments. 🤔 

Would be good to add an extra test case.

Maybe we need to write `flatten_but_keep_block` instead of `flatten_args`?  Something like `flatten_but_keep_block (x) = (y = flatten(x); isexpr(x, :block) && !isexpr(y, :block) ? Expr(:block, y) : y)`. Or maybe there's a less wasteful implementation.

Reply:
yeah, I am leaning in the same direction on all of your suggestions. I will plan to send you something over the weekend.